### ISSUE TITLE
Improvements to container atributes

### ISSUE THREAD

**Issue by samuelsleight** [sharksforarms/deku#125]
A bit of a vague title, but it's a bit of a vague, open-ended issue:

We currently have some inconsistencies when it comes to attributes on `Vec` fields that feel like they could lead to some confusion or result in things being unexpectedly impossible. One example of this is the `bits` (or `bytes`) attribute: on all non-`Vec` fields, this is the number of bits to read for the entire field, but for `Vec` fields it is the number of bits to read for each item in the `Vec` - the number of bits for the entire field is a different attribute entirely (pending #124), which seems potentially unexpected. For _extra_ confusion, it is _not_ the case for _all_ attributes that they suddenly apply on a per-element basis: take `map`, for example, which is applied to the entire `Vec` field - there is no way of applying a `map` to each element.

It would be interesting to contemplate some ideas for resolving this - I've thought of a couple, but I don't know if I'm sold by either of them, (or indeed whether the second is particularly feasible, as I have limited experience with proc-macros):

The first, which led me down this train of thought but otherwise I'm not a fan of, is to potentially take a comma-separated (or similar) list of parameters for each attribute on a `Vec, like the following (admittedly convoluted) examples:
```rust
#[deku(bits="15, 5")]
vec_a: Vec<u8>,

#[deku(bytes="2, _", map="_, |value: u8| Ok(value + 1)"
vec_b: Vec<u8>
```

`vec_a` would read 15 bits worth of data, taking 5 bits at a time for each `u8` added to the container. `vec_b`would read 2 bytes of data, and for each read value apply the mapping function to add 1 - the `_`s here mean we are not passing a value for this attribute at that level.

A second, _marginally_ nicer idea, could be to use two lists of attributes for a `Vec` field, if that is something that is actually possible:
```rust
#[deku(bits="15")]
#[deku(bits="5")]
vec_a: Vec<u8>,


#[deku(bytes="2")]
#[deku(map="|value: u8| Ok(value + 1)")]
vec_b: Vec<u8>
```

This would be equivalent to the first example.

As mentioned I don't know if I'm particularly sold by either of those suggestions, but hopefully they serve to get the idea across and spark some potentially interesting discussion

**Comment by sharksforarms** [comment]
Thanks for creating this issue! I agree with what you've said, currently it's hard to know which attributes apply to the container/value inside the container, and there's not much control as to where they apply.

Another idea:
```rust
#[deku(
    outer(count = "3"),
    inner(bits = "3")
)]
vec: Vec<u8>
```

**Comment by samuelsleight** [comment]
Ooh, yeah, I certainly think that's a nicer syntax that my suggestions.

One other thing to think about is potentially making this extensible to nested containers - i.e. what syntax would work were we to support `Vec<Vec<u8>>`. That's probably a very unlikely case, but I wouldn't like to rule it out.

**Comment by sharksforarms** [comment]
> One other thing to think about is potentially making this extensible to nested containers - i.e. what syntax would work were we to support `Vec<Vec<u8>>`. That's probably a very unlikely case, but I wouldn't like to rule it out.

Yeah, it would be a good idea to have something which doesn't assume two...

(Assuming this is possible..)
```rust
#[deku(
    count = "3", 
    deku(
        count = "1",
        deku(bits = "3")
    )
)]
```

**Comment by samuelsleight** [comment]
I _think_ I preferred the use of `outer` and `inner`, just for them being slightly more explicit as to their meaning, though perhaps we could just drop the `outer` and have a chain of `inner`s instead:
```rust
#[deku(
    count = "3", 
    inner(
        count = "1",
        inner(bits = "3")
    )
)]
```

**Comment by samuelsleight** [comment]
Having written that, I remembered about #129, which potentially throws another spanner in the works - we could perhaps replace `inner` with `key` and `value`, where `key` is relevant for `HashMaps` or similar and is relevant when reading the key values, and `value` is always relevant and is for reading items in a `Vec` or the values being keyed in a `HashMap`?

**Comment by sharksforarms** [comment]
inner does _look_ the best so far.

```rust
#[deku(
    count = "3", 
    value(
        count = "1",
        value(bits = "3")
    )
)]
```

**Comment by samuelsleight** [comment]
Yeah, I think I like that the best out of what we have so far. We could potentially have `inner` and `value` be interchangable, or maybe `key` and `inner` is fine if we prefer `inner` to `value`?

**Comment by CJKay** [comment]
How about `each`? It's the first thing to my mind - "for each element do X".

```rust
#[deku(count = "3", each(count = "1", each(bits = "3"))]
```
